src/data/set_like/basic.lean

```diff
@@ -160,4 +160,14 @@ lemma exists_of_lt : p < q → ∃ x ∈ q, x ∉ p := set.exists_of_ssubset
 lemma lt_iff_le_and_exists : p < q ↔ p ≤ q ∧ ∃ x ∈ q, x ∉ p :=
 by rw [lt_iff_le_not_le, not_le_iff_exists]
 
+/-- If a dependent function acts on a indexed collection of subtypes, then when the indices
+are the same the subtype argument can be rewritten.
+
+See [this Zulip thred](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/simp.20simp/near/263872040)
+for dome discussion about this lemma. It can occasionally result in `simp, simp` being needed. -/
+@[congr] lemma dep_congr {ι α} [set_like A B]
+  {p : ι → A} (f : Π i, p i → α) :
+  ∀ {i j} (x : p i) (h : i = j), f i x = f j ⟨x, h ▸ x.prop⟩
```

Comment:
As discussed in that Zulip thread, this should not be a congr lemma because it 1) does higher-order matching, 2) it does it on every subterm (since the head symbol is a variable and thus indexing bails out), and 3) it can't rewrite at x.

Reply:
I'm curious to see whether mathlib builds anyway.

Reply:
If I understand @gebner correctly, this is not the correct solution. Instead, core should be fixed. But I don't know if anyone wants to pour energy in that.

Reply:
Yes, that's my understanding too. It would be great if we could check if lean4 core works for this type of case, if we decide no one has the energy to make it work in lean3